erc777.k

```diff
@@ -69,140 +72,379 @@ module ERC777
+         ...
+       </account>
+       <account>
+         <id> To </id>
+         <balance> ValueTo => ValueTo +Int Value </balance>
+         ...
+       </account>
        requires From =/=Int To
         andBool ValueFrom -Int Value >=Int 0
-        andBool ValueTo   +Int Value  <Int #maxNumber
+        andBool ValueTo   +Int Value  <=Int #MAXVALUE
 
   rule <k> #update ( From , To , Value ) => . ... </k>
        requires From ==Int To
 
-  syntax Stmt ::= #log ( Address , Address , Address , Number , Data , Data )
+  syntax BExp ::= #isMultipleOf ( AExp , AExp ) [seqstrict]
+
+  rule #isMultipleOf ( Value:Int , Granularity:Int ) => Value /Int Granularity *Int Granularity ==Int Value
+
+  //
+  // burn / operatorBurn
+  // TODO: inconsistencies in the standards and reference implementation
+
+  // NOTE: contract MAY prevent some/all holders from burning for any reason
+  // vs Tokens MAY be burned from any holder address
 
-  rule <k> #log ( Operator , From , To , Value , Data , OperatorData ) => . ... </k>
-       <log> Log => Log Sent ( Operator , From , To , Value , Data , OperatorData ) </log>
+  rule <k>         burn(        Amount:Int)
+        => operatorBurn(Holder, Amount, "") ... </k>
+       <caller> Holder </caller>
 
-  syntax Exp ::= #isMultipleOf ( Number , Number )
+  rule <k> operatorBurn(Holder:Int, Amount:Int, OperatorData)
+        => #atomicBegin
+        ~> #ensure(#and(isOperatorFor(Operator, Holder),
+                        #isMultipleOf(Balance -Int Amount, granularity()),
+                        #and(Amount >=Int 0, Balance >=Int Amount),
+                        Holder =/=Int 0))
+        ~> #burn(Holder, Amount)
+        ~> #callRecipent(Operator, Holder, 0, Amount, "", OperatorData, true)
+        ~> #log(Burned(Operator, Holder, Amount, OperatorData))
+        ~> #logERC20(Transfer(Holder, 0, Amount)) // ERC20
+        ~> #atomicEnd(#void) ... </k>
+       <caller> Operator </caller>
+       <id> Holder </id>
+       <balance> Balance </balance>
+
+  syntax Stmt ::= #burn ( Address , Number )
+
+  rule <k> #burn(Holder, Amount) => . ... </k>
+       <account>
+         <id> Holder </id>
+         <balance> Balance => Balance -Int Amount </balance>
+         ...
+       </account>
+       <supply> Supply => Supply -Int Amount </supply>
+       requires Balance >=Int Amount andBool Supply >=Int Amount
+
+  // general version
+  syntax Stmt ::= #log ( Event )
+                | #logERC20 ( Event )
+
+  rule <k> #log(Log) => . ... </k>
+       <log> Logs => Logs Log </log>
+
+  rule <k> #logERC20(Log) => . ... </k>
+       <log> Logs => Logs Log </log>
+       <ERC20CompatibleC> ERC20Compatible </ERC20CompatibleC>
+       requires ERC20Compatible
+
+  rule <k> #logERC20(Log) => . ... </k>
+       <ERC20CompatibleC> ERC20Compatible </ERC20CompatibleC>
+       requires notBool ERC20Compatible
+
+
+  //
+  // mint -> no function spec, but rules
+  // TODO: `event Minted` signature inconsistency in standard
```

Comment:
we need to make it clear that `mint` is not part of the standard, and this is an example of the implementation. indeed, it's all about the testing for us.